test/performance/odp_ml_perf.c

```diff
@@ -0,0 +1,1051 @@
+		ret = -1;
+		goto error;
+	}
+
+	memcpy(input, glb->inp_file_data, glb->inp_size_q);
+	inp_addr = input;
+
+	if (glb->opt.scale_q > 0.0) {
+		input = malloc(glb->inp_size_q);
+		if (!input) {
+			ODPH_ERR("Allocating %" PRIu64 " bytes failed\n", glb->inp_size_q);
+			ret = -1;
+			goto error;
+		}
+
+		inp_addr = input;
+	}
+
+	for (int i = 0; i < glb->num_inp; i++) {
+		inp_seg[i].addr = inp_addr;
+		inp_seg[i].size = glb->inp[i].size;
+		inp_addr += glb->inp[i].size;
+	}
+
+	output = malloc(glb->out_size_q);
+
+	if (!output) {
+		ODPH_ERR("Allocating %" PRIu64 " bytes failed\n", glb->out_size_q);
+		ret = -1;
+		goto error;
+	}
+
+	out_addr = output;
+	output_final = output;
+	out_size_final = glb->out_size_q;
+
+	if (glb->opt.scale_d > 0.0) {
+		output_file = malloc(glb->out_size_d);
+		if (!output_file) {
+			ODPH_ERR("Allocating %" PRIu64 " bytes failed\n", glb->out_size_d);
+			ret = -1;
+			goto error;
+		}
+
+		output_final = output_file;
+		out_size_final = glb->out_size_d;
+	}
+
+	for (int i = 0; i < glb->num_out; i++) {
+		out_seg[i].addr = out_addr;
+		out_seg[i].size = glb->out[i].size;
+		out_addr += glb->out[i].size;
+	}
+
+	odp_ml_data_t data = {
+		.input_seg = inp_seg,
+		.num_input_seg = glb->num_inp,
+		.output_seg = out_seg,
+		.num_output_seg = glb->num_out,
+	};
+	odp_ml_run_param_t run_param;
+
+	odp_ml_run_param_init(&run_param);
+	run_param.batch_size = glb->opt.num_batch;
+
+	if (glb->opt.mode != MODE_INFERENCE_QUANT && glb->opt.scale_q > 0.0) {
+		if (quantize_input(input, glb->inp_file_data)) {
+			ret = -1;
+			goto error;
+		}
+	}
+
+	odp_barrier_wait(&glb->barrier);
+
+	for (int i = 0; i < glb->opt.rounds + glb->opt.warmup; i++) {
+		if (i == glb->opt.warmup)
+			time_start(&time);
+
+		if (glb->opt.mode == MODE_INFERENCE_QUANT && glb->opt.scale_q > 0.0) {
+			if (quantize_input(input, glb->inp_file_data)) {
+				ret = -1;
+				goto error;
+			}
+		}
+
+		int r;
+
+		while (!(r = odp_ml_run(glb->mdl, &data, &run_param)))
+			;
+		if (r != 1) {
+			ODPH_ERR("odp_ml_run() failed\n");
+			ret = -1;
+			goto error;
+		}
+
+		if (glb->opt.mode == MODE_INFERENCE_QUANT && glb->opt.scale_d > 0.0) {
+			if (dequantize_output(output_file, output)) {
+				ret = -1;
+				goto error;
+			}
+		}
+
+		if (glb->opt.latency && i >= glb->opt.warmup)
+			time_elapsed(&time, thread_idx);
+	}
+
+	if (!glb->opt.latency)
+		time_elapsed(&time, thread_idx);
+
+	if (glb->opt.mode != MODE_INFERENCE_QUANT && glb->opt.scale_d > 0.0) {
+		if (dequantize_output(output_file, output)) {
+			ret = -1;
+			goto error;
+		}
+	}
+
+	if (glb->ref_file_data) {
+		if (out_size_final != glb->ref_file_size) {
+			ODPH_ERR("Output size mismatch: %" PRIu64
+				 " differs from reference file size %" PRIu64 "\n",
+				 out_size_final, glb->ref_file_size);
+			ret = -1;
+			goto error;
+		}
+
+		if (memcmp(glb->ref_file_data, output_final, out_size_final)) {
+			ODPH_ERR("Output differs from reference\n");
+			ret = -1;
+			goto error;
+		}
+	}
+
```

Comment:
Some barrier at the end likely needed so that stats are up-to-date when threads finish.

Reply:
Thread join is that barrier.